PartDesign: Fillet, Chamfer, Draft, Thickness rework.

PaddleStroke opened the pull request:
For the tools Fillet, Chamfer, Draft, Thickness in PartDesign : 
- Merge Add and Remove in one button.
- Enable those features to be empty. This let us start the tool before selection.
- Refactor the code that had needless duplicates.
- Fix bug (fillet, click addAllEdges after having a face in the list. Then the list has more entries than the feature has references. So if you delete the last element of the list, crash.)

Issue : https://github.com/FreeCAD/FreeCAD/issues/8909

Forum Dicussion :
https://forum.freecad.org/viewtopic.php?t=76989&start=10

Video explanation : 
https://youtu.be/3mHUwT8-pwQ

Testers and reviewers welcome !


Thank you for creating a pull request to contribute to FreeCAD! Place an "X" in between the brackets below to "check off" to confirm that you have satisfied the requirement, or ask for help in the [FreeCAD forum](https://forum.freecadweb.org/viewforum.php?f=10) if there is something you don't understand.

- [x]  Your Pull Request meets the requirements outlined in section 5 of [CONTRIBUTING.md](https://github.com/FreeCAD/FreeCAD/blob/master/CONTRIBUTING.md) for a Valid PR

Please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [1.0 Changelog Forum Thread](https://forum.freecad.org/viewtopic.php?f=10&t=69438).

---

luzpaz commented:
@PaddleStroke youtube clip is in private mode. Can you change it to unlisted ?

PaddleStroke commented:
It should work now @luzpaz

luzpaz commented:
Nice UX refinements! Have you checked if the Undo/Redo functionality is affected with these new refinements?

PaddleStroke commented:
I haven't noticed anything peculiar, why would you ask ? Do you expect some specific way for it to be affected?

adrianinsaval commented:
very nice improvement, however I see that you can make a fillet with no reference selected. I think it's good to start the tool with no selection or to allow empty selection during editing, but IMO once clicking ok an empty fillet should at least raise a warning or even an error.

PaddleStroke commented:
Maybe we can set the OK button to complain and not proceed if the list is empty.

howie-j commented:
Much love to @PaddleStroke for this improvement, a million times better than the current solution!

> Maybe we can set the OK button to complain and not proceed if the list is empty.

Is it possible to just delete/remove the fillet from the work tree if nothing is selected? (lets not add unnecessary warnings/errors, thats just annoying :))

adrianinsaval commented:
if your intention is to not make a fillet use the cancel button.

howie-j commented:
Yes but pressing enter is faster and easier.

However thinking about it a bit more, what about this:
- Delete/remove fillet from the feature tree
- Show a short/non-intrusive warning (the recent feature from abdullah) that says something like this: "Nothing selected, no fillet created", and disappears after a short time

PaddleStroke commented:
The issue with OK deleting the empty fillet is that if a user is editing a fillet, removes all edges and click OK. Then the fillet object gets deleted and what if the fillet was in the middle of the tree? He can't put it back there.
Maybe OK button can act the same as 'Cancel' button if it's empty.

But are we sure we want to prevent empty fillets ? Why the user would validate an empty fillet? I mean this is creating a limitation that is probably not necessary. Maybe the user wants to create an empty fillet that he will fill later (though with tnp that wouldn't be a good idea ! :D ).

I think my prefered solution would be to be as little intrusive as possible. So not adding limitation or deleting the fillet. Only a non-intrusive Console().Warning()

adrianinsaval commented:
I would like empty fillets to have some form of warning icon or similar in the tree

freecadci commented:
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8990"><img alt="pipeline status" src="https://gitlab.com/freecad/FreeCAD-CI/badges/PR_8990/pipeline.svg"/></a> for feature branch [PR_8990](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8990). Pipeline [817026063](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/817026063) was triggered at [2e250af](https://github.com/FreeCAD/FreeCAD/pull/8990/commits/2e250af9bc35eb3e22d4b558eadb7218e1b4c93e). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

sliptonic commented:
Rather than labeling the button as 'start selection' and 'end selection'  make it relevant to the user's intent.  Something like 'Select' and 'Preview'

donovaly commented:
many thanks!
@PaddleStroke , can you please rebase (compilation of the PR takes ages)

What I miss is the unification with the Part WB. As we spoke, we have duplicated code in part and Part WB and it should be merged to one file in Part that is used by Part and PD like we did for Pad/Pocket.

donovaly commented:
> What I miss is the unification with the Part WB. As we spoke, we have duplicated code in part and Part WB and it should be merged to one file in Part that is used by Part and PD like we did for Pad/Pocket.

OK, I see that you first did the UI thing, so let's postpone the merge/revision wok to another PR.

Concerning the new workflow of this PR, what should I say? I love it?
Yes, I love it ;-) Well done!

> Rather than labeling the button as 'start selection' and 'end selection' make it relevant to the user's intent. Something like 'Select' and 'Preview'

I agree, this would me more logical since once the selection mode is turned off you get a preview.

PaddleStroke commented:
Merging of the tools is also a good idea that I think we should pursue. I can work on that later. If we do so, tools should be in Part by default not PartDesign?

Here I rebased and changed the strings to 'Preview' and 'Select'.

freecadci commented:
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8990"><img alt="pipeline status" src="https://gitlab.com/freecad/FreeCAD-CI/badges/PR_8990/pipeline.svg"/></a> for feature branch [PR_8990](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8990). Pipeline [817639688](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/817639688) was triggered at [5c996b0](https://github.com/FreeCAD/FreeCAD/pull/8990/commits/5c996b072155beaa64f53876380baaee7799ed26). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

PaddleStroke commented:
The CI is complaining, but the report talk about FEM test failed. So I think it's because I rebased on master and something merged in FEM that broke the test?

donovaly commented:
> The CI is complaining, but the report talk about FEM test failed. So I think it's because I rebased on master and something merged in FEM that broke the test?

Yes, I somehow managed to break my PR. I reverted. When you rebase and force-push, the CI here should succeed again. Sorry for the inconvenience.

0penBrain commented:
There are some consistency issues. The button is sometimes initialized with "Enter selection", sometimes with "Start selection".
Then once used, it toggles between "Preview" and "Select" only, initial value is never used again.

PaddleStroke commented:
Thanks @0penBrain I will fix this.

chennes commented on src/Mod/PartDesign/App/FeatureThickness.cpp:
```diff
@@ -72,8 +72,22 @@ App::DocumentObjectExecReturn *Thickness::execute()
+    }
+
+    /* If the feature was empty at some point, then Placement was set by positionByBaseFeature.
+    *  However makeThickSolid apparantly requires the placement to be empty, so we have to clear it*/
```
```suggestion
    *  However makeThickSolid apparently requires the placement to be empty, so we have to clear it*/
```

chennes commented on src/Mod/PartDesign/Gui/Command.cpp:
```diff
@@ -1541,7 +1541,7 @@ bool CmdPartDesignSubtractiveHelix::isActive()
 
 bool dressupGetSelected(Gui::Command* cmd, const std::string& which,
-        Gui::SelectionObject &selected, bool &useAllEdges)
+        Gui::SelectionObject &selected, bool &useAllEdges, bool& noSelection)
```
I'm a bit concerned about "boolean blindness" here -- we've got a function that effectively returns _three_ booleans now. Consider filing this away as a potential spot for a future refactoring.

chennes commented on src/Mod/PartDesign/Gui/Command.cpp:
```diff
@@ -1804,23 +1815,32 @@ void CmdPartDesignThickness::activated(int iMsg)
+        size_t i = 0;
+        while (i < SubNames.size())
+        {
+            std::string aSubName = static_cast<std::string>(SubNames.at(i));
```
I know it's not really your code, but why is this `static_cast` here? Doesn't `at()` give you a std:string??

PaddleStroke replied:
Yes it does return a std::string. I don't know why this was here.

freecadci commented:
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8990"><img alt="pipeline status" src="https://gitlab.com/freecad/FreeCAD-CI/badges/PR_8990/pipeline.svg"/></a> for feature branch [PR_8990](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8990). Pipeline [820233711](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/820233711) was triggered at [1258815](https://github.com/FreeCAD/FreeCAD/pull/8990/commits/12588150c1117e8bfcdeb3cb5076ac99e48e31f0). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

PaddleStroke commented:
I just fixed your different comments. Please let me know if you find something else.

chennes commented:
Love it, thanks. Let's get this merged, and then if there's more follow-up work to be done it can go in a separate PR.

luzpaz commented:
Excellent work and collaboration!
